src/ark/utils/plot_utils.py

```diff
@@ -147,14 +144,30 @@ def _metacluster_cmap_generator(self) -> None:
             f"{self.cluster_type}_meta_cluster"
         ].map(self.metacluster_colors)
 
-        # Convert the list of tuples to a numpy array, each index is a color
+        # grab mask cluster_id integers and merge with raw_cmap
+        cluster_id_to_metacluster_map = cluster_id_to_name[
+            [f"{self.cluster_type}_meta_cluster", "cluster_id"]].drop_duplicates()
+        unassigned_cluster_id: int = int(cluster_id_to_name["cluster_id"].max() + 1)
+        cluster_id_to_metacluster_map = pd.concat(
+            [
+                cluster_id_to_metacluster_map,
+                pd.DataFrame(
+                    data={
+                        f"{self.cluster_type}_meta_cluster": [unassigned_cluster_id, 0],
+                        "cluster_id": [unassigned_id, 0],
+                    }
+                )
+            ]
+        )
+        cluster_id_to_color = metacluster_id_to_name.merge(
+            cluster_id_to_metacluster_map, on=[f"{self.cluster_type}_meta_cluster"])
 
-        mc_colors: np.ndarray = np.array(metacluster_id_to_name['color'].to_list())
+        # sort by cluster_id ascending, so colors align with mask integers
+        cluster_id_to_color.sort_values(by="cluster_id", inplace=True)
+        cluster_id_to_color.reset_index(drop=True, inplace=True)
 
-        metacluster_to_index = {}
-        metacluster_id_to_name.reset_index(drop=True, inplace=True)
-        for index, row in metacluster_id_to_name.reset_index(drop=True).iterrows():
-            metacluster_to_index[row[f'{self.cluster_type}_meta_cluster']] = index
+        # Convert the list of tuples to a numpy array, each index is a color
+        mc_colors: np.ndarray = np.array(cluster_id_to_color['color'].to_list())
```

Comment:
I think everything above this is a little bit needlessly confusing. It seems like all of this code is to first map metacluster to color, and then metacluster to cluster_id, then cluster_id to color while adding an unassigned id twice. I think it can be simplified to add the unassigned id only once.

I think this is much simpler, starting from line 106 (sorry didn't have time to test it so might have issues, but this is the gist of what I'm thinking, let me know if there's a flaw my logic):

```
        # subset on metacluster, metacluster name, and cluster id
        metacluster_id_to_name = cluster_id_to_name[
            [f"{self.cluster_type}_meta_cluster", f"{self.cluster_type}_meta_cluster_rename","cluster_id"]
        ].copy()
        unassigned_meta_cluster: int = int(
            metacluster_id_to_name[f"{self.cluster_type}_meta_cluster"].max() + 1)
        unassigned_cluster_id: int = int(
            metacluster_id_to_name["cluster_id"].max() + 1)

        # Extract unique (metacluster-ID,  name, cluster_id)
        # Set the unassigned meta cluster to be the max ID + 1
        # Set 0 as the Empty value
        metacluster_id_to_name: pd.DataFrame = pd.concat(
            [
                metacluster_id_to_name.drop_duplicates(),
                pd.DataFrame(
                    data={
                        f"{self.cluster_type}_meta_cluster": [unassigned_meta_cluster, 0],
                        f"{self.cluster_type}_meta_cluster_rename": ["Unassigned", "Empty"],
                        "cluster_id":[unassigned_cluster_id, 0]
                    }
                )
            ]
        )

        # add the unassigned color to the metacluster_colors dict
        self.metacluster_colors.update({unassigned_meta_cluster: self.unassigned_color})

        # add the no cluster color to the metacluster_colors dict
        self.metacluster_colors.update({0: self.background_color})

        # assert the metacluster index in colors matches with the ids in metacluster_id_to_name
        misc_utils.verify_same_elements(
            metacluster_colors_ids=list(
                self.metacluster_colors.keys()),
            metacluster_mapping_ids=metacluster_id_to_name
            [f'{self.cluster_type}_meta_cluster'].values)
        # use metacluster_colors to add the colors to metacluster_id_to_name
        metacluster_id_to_name["color"] = metacluster_id_to_name[
            f"{self.cluster_type}_meta_cluster"
        ].map(self.metacluster_colors)

        # sort by cluster_id ascending, so colors align with mask integers
        metacluster_id_to_name.sort_values(by="cluster_id", inplace=True)
        metacluster_id_to_name.reset_index(drop=True, inplace=True)

        # Convert the list of tuples to a numpy array, each index is a color
        mc_colors: np.ndarray = np.array(metacluster_id_to_name['color'].to_list())
```

Reply:
Yeah this makes sense to me and should work! I was trying to avoid pd.merge() issues with the two dfs but just adding cluster_id to the original metacluster_id_to_name totally solves that!